Enforce difficulty setting limits during decode

Hecatia-Lapislazuli opened the pull request:
Closes #27172

To address #27172, adding a minimum value of 0.02 for the multiplier for circle size. 0.02 was chosen as it appears to be the smallest value (or well, somewhere between 0.01 and 0.02) that one can have without a crash.

Alternatively, could be capped to 0.15, that is a CS of 10, which should be the highest normally occurring, however feel is better to cap it at the smallest value possible before it's effect becomes meaningless to allow for as much freedom in design of beatmaps as possible.

Hecatia-Lapislazuli committed:
Added a minimum value for the scale calculated by the CS value.

bdach commented:
If this is going to be enforced, I would argue it should be enforced in the beatmap decoder, not at point of usage. See: 

https://github.com/ppy/osu/blob/9532725eb78e05b70bb89842330553a8a95514ef/osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs#L405

or:

https://github.com/ppy/osu/blob/9532725eb78e05b70bb89842330553a8a95514ef/osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs#L409

And the limits should probably be [0;10] rather than the twelve point whatever at which point scale goes negative.

Rationale: having it enforced at point of usage causes weirdness because you can put circle size 555 and it'll be shown on UI everywhere as that, but in reality it'll work the same as circle size twelve point whatever. Clamping in the decoder just throws away such undefined behaviour (you can't get to CS12 via any standard means other than editing beatmaps, so I consider it undefined behaviour).

@smoogipoo thoughts?

Hecatia-Lapislazuli commented:
Restricting CS in the beatmap decoder currently breaks at least one ranked map ranked beatmap [4101927](https://osu.ppy.sh/beatmapsets/1976303#mania/4101927) - albeit for mania. Should be possible to make an exception for mania there, I suppose, but it isn't ideal.

But I suppose fair enough for non-mania to be limited, just from my perspective as while yes, it does require editing of the beatmap file, it can be interesting to have yet for gimmicks.

I'd also still personally prefer it to occur in the method where you'd expect the result to always be a positive real number, and also makes the fix a bit more robust to changes in the future as well (as an example, the above issue with mania).

bdach commented:
Fair point on mania, I forgot 10+k is now rankable there which makes this more difficult to enforce. But I'd still want to restrict all of the other rulesets to a known good limit.

What you call "gimmicks" I will continue to call "undefined behaviour" and the amount of undefined behaviour from stable that we have to continue to support is what requires lazer code to be more complex than I'd personally ideally expect. So I still am in favour of drawing very clear, very explicit red lines.

Hecatia-Lapislazuli commented:
Hmm... Well, just difference in philosophy I guess, but I'm really not someone in a position to make a subjective decision like that either. 

And I'm somewhat more agreeing with you now as have since looked into it. I thought there were a fair few loved maps that had hp/od/ar being greater than 10 which i felt was precedent enough to leave it be, but I can only see one loved map that satisfies this [4243833](https://osu.ppy.sh/beatmapsets/1263883#osu/4243833) with an AR of between 10 and 11, and hardly any in the graveyard either

Regardless, I'd still prefer the scale method to also have a limit regardless of if the input is restricted, just as the fact the method can return a value that causes crashes is in and of itself an issue.

But for now, I'll write up something later today/tomorrow to implement restricting the beatmap decoding to CS≤10. 

Also something I suppose to consider, should HP/OD/AR be restricted to be equal to or bellow 10? Given there's currently only one loved map that this would cause issues with... can perhaps be better to consider doing now rather then later.
Plus given this would only be lowering the AR of said map, it shouldn't cause issues with replays either.

bdach commented:
> Also something I suppose to consider, should HP/OD/AR be restricted to be equal to or bellow 10?

[0;10] is the range we expose on ui. So it should be that or nothing.

Hecatia-Lapislazuli committed:
Clamped Difficulty Ranges to [0,10]

Hecatia-Lapislazuli commented:
Okay, here's a commit for this, with additional clamps for HP/OD/AR.

Let me know if anything else should be changed.

Hecatia-Lapislazuli committed:
Moved DIfficulty Clamping to occur after the file has been parsed
This is to handle potential issues with the ruleset being parsed after circle size has been parsed.

Hecatia-Lapislazuli commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -383,21 +383,24 @@ private void handleDifficulty(string line)
                 case @"CircleSize":
                     difficulty.CircleSize = Parsing.ParseFloat(pair.Value);
+                    //If the mode is not Mania, clamp circle size to [0,10]
+                    if (!beatmap.BeatmapInfo.Ruleset.OnlineID.Equals(3))
```
Hmm... I did think of this, but sample I took had it being fine... But i suppose even if my sample was fine, that doesn't mean the user ain't gonna do something funky with editing the file.

I just wanted to avoid having to do checks in `LegacyBeatmapDecoder.ParseStreamInto`, originally my plan was just to change all the `BeatmapDifficulty` getters to do this, but it lead to way more complications then it was worth. (breaking stuff with realms, plus certain mods allow the user to edit difficulty settings up to 11, and the fact that you can't access what ruleset is being used from this class which required reworking how Mania checks circle size)

Anywho, latest commit should address this. Also added a clamp for when the mode is Mania - not as needed I suppose, but I think is good for consistency.

smoogipoo commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -102,6 +104,26 @@ protected override void ParseStreamInto(LineBufferedReader stream, Beatmap beatm
+            //If it is Mania, it must be within [1,20] - dual stages with 10 keys each.
+            //The lower bound should be 4, but there are ranked maps that are lower than this.
+            else
+                difficulty.CircleSize = Math.Clamp(difficulty.CircleSize, 1, 20);
```
Technically only [1,18] is supported. Was 20 an intentional choice?

Hecatia-Lapislazuli replied:
May've misread the code in that case - we have `ManiaRuleset.MAX_STAGE_KEYS = 10`, so I assumed that the limit was 20... But ah, I assume allowing when single to have 10 keys, but only allowing 9 when dual would make sense.
I suppose out of scope, but should `for (int i = 2; i <= MAX_STAGE_KEYS * 2; i += 2)` be changed to `for (int i = 2; i <= (MAX_STAGE_KEYS-1) * 2; i += 2)` in that case? 

https://github.com/ppy/osu/blob/3a608aa8153541f4f610d7bea018ce67586f9907/osu.Game.Rulesets.Mania/ManiaRuleset.cs#L327-L336

bdach replied:
Stable does this:

```csharp
                                        case "CircleSize":
                                            if (BeatmapVersion >= 13)
                                            {
                                                if (PlayMode == PlayModes.OsuMania)
                                                    DifficultyCircleSize = Math.Min(18.0f, Math.Max(1.0f, Single.Parse(right, GameBase.nfi)));
                                                else
                                                    DifficultyCircleSize = Math.Min(10.0f, Math.Max(0.0f, Single.Parse(right, GameBase.nfi)));
                                            }
                                            else
                                                DifficultyCircleSize = Math.Min((byte)10, Math.Max((byte)0, byte.Parse(right)));
```

https://github.com/peppy/osu-stable-reference/blob/0e425c0d525ef21353c8293c235cc0621d28338b/osu!/GameplayElements/Beatmaps/Beatmap.cs#L577-L586

The 18 thing may be coming from [ranking criteria](https://osu.ppy.sh/wiki/en/Ranking_criteria/osu%21mania):

>  Beatmaps can only use 4–10, 12, 14, 16, or 18 keys.

I don't see much harm in supporting 20, albeit there is an argument we should possibly be disallowing odd keycounts above 10 if we're to match rc. Not sure.

smoogipoo replied:
I believe osu!stable supports odd key counts but not >18K.

Without getting into the weeds about our stance on backwards compatibility and all that, I think we take the easiest path forward and emulate exactly what stable's doing.

bdach replied:
> Hmm, I receive 404 for the github link...

Yes that's a github link that is supposed to 404 for anyone that's not in the core team.

Hecatia-Lapislazuli replied:
Ahhh, okay. I thought at first it took me to the actual page, so was a bit confused.

smoogipoo commented on osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:
```diff
@@ -102,6 +104,26 @@ protected override void ParseStreamInto(LineBufferedReader stream, Beatmap beatm
+        {
+            difficulty.DrainRate = Math.Clamp(difficulty.DrainRate, 0, 10);
+            //If the mode is not Mania, clamp circle size to [0,10]
+            if (!beatmap.BeatmapInfo.Ruleset.OnlineID.Equals(3))
```
Use `== 3` or `!= 3`.

bdach commented on osu.Game/Rulesets/Objects/Legacy/LegacyRulesetExtensions.cs:
```diff
@@ -55,7 +55,7 @@ public static float CalculateScaleFromCircleSize(float circleSize, bool applyFud
             const float broken_gamefield_rounding_allowance = 1.00041f;
 
-            return (float)(1.0f - 0.7f * IBeatmapDifficultyInfo.DifficultyRange(circleSize)) / 2 * (applyFudge ? broken_gamefield_rounding_allowance : 1);
+            return (float)Math.Max(0.02, (1.0f - 0.7f * IBeatmapDifficultyInfo.DifficultyRange(circleSize)) / 2 * (applyFudge ? broken_gamefield_rounding_allowance : 1));
```
FWIW I still think this should be gone from this diff and only the parser change should remain.

Hecatia-Lapislazuli replied:
I do think a check of this sort is useful to prevent the function from returning a result out of range, but sure. I've reverted it in the latest commit.

bdach reviewed (commented):
Tests for this would be nice.

Hecatia-Lapislazuli committed:
Changed allowed mania keys, and reverted 0af32c5

Hecatia-Lapislazuli committed:
Added tests for difficulty clamping

Hecatia-Lapislazuli commented:
I believe this should work for tests - although I haven't used them before, so not too certain. Advice would be appreciated if anything needs improvements/should have been done differently.

bdach committed:
Menial cleanups

smoogipoo committed:
Add whitespace to appease R#

peppy committed:
Make some more methods static
